File: src/main/scala/com/zotoh/frwk/util/CmdLineSeq.scala
/*??
 * COPYRIGHT (C) 2012 CHERIMOIA LLC. ALL RIGHTS RESERVED.
 *
 * THIS IS FREE SOFTWARE; YOU CAN REDISTRIBUTE IT AND/OR
 * MODIFY IT UNDER THE TERMS OF THE APACHE LICENSE,
 * VERSION 2.0 (THE "LICENSE").
 *
 * THIS LIBRARY IS DISTRIBUTED IN THE HOPE THAT IT WILL BE USEFUL,
 * BUT WITHOUT ANY WARRANTY; WITHOUT EVEN THE IMPLIED WARRANTY OF
 * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.
 *
 * SEE THE LICENSE FOR THE SPECIFIC LANGUAGE GOVERNING PERMISSIONS
 * AND LIMITATIONS UNDER THE LICENSE.
 *
 * You should have received a copy of the Apache License
 * along with this distribution; if not, you may obtain a copy of the
 * License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 ??*/

package com.zotoh.frwk
package util

import scala.collection.mutable

import java.io.{BufferedOutputStream,InputStreamReader,OutputStreamWriter}
import java.io.{Reader,Writer}
import java.util.{Properties=>JPS}
import org.apache.commons.lang3.{StringUtils=>STU}
import com.zotoh.frwk.util.StrUtils._
import com.zotoh.frwk.util.CoreUtils._


/**
 * @author kenl
 *
 */
abstract class CmdLineSeq( qs:CmdLineQ* ) {

  private var _out= new OutputStreamWriter( new BufferedOutputStream(System.out))
  private var _in = new InputStreamReader(System.in)
  private val _Qs= mutable.HashMap[String,CmdLineQ]()
  private var _par:CmdLineSeq= null
  private var _canceled=false

  qs.foreach { (q) =>  _Qs +=  q.label() -> q  }

  /**
   * @param props
   */
  def start(props:JPS) : this.type = {
    val kp= if (isWindows()) "<Ctrl-C>" else "<Ctrl-D>"
    var stop= false
    if (_par == null) {
//        println( ">>> Press <Ctrl-D><Enter> to cancel...")
        println(  ">>> Press "+kp+"<Enter> to cancel...")
    } else if ( _par.start(props).isCanceled) {
        _canceled=true
        end()
        stop=true
    }
    if (!stop) {
      _Qs.get(onStart) match {
        case Some(c) => cycle(c, props)
        case _ => end()
      }
    }
    this
  }

  /**
   * @return
   */
  def isCanceled() = _canceled

  /**
   * @param id
   */
  def remove(id:String): this.type = { _Qs -= id ; this }

  private def cycle(original:CmdLineQ, props:JPS) {
    var c:CmdLineQ =original
    var n=""
    while (c != null) {
      c.setOutput(props)
      n=popQuestion(c)
      c = if (n == "") null else {
       if (_Qs.isDefinedAt(n)) _Qs( n) else null
      }
    }
    end()
  }

  /**
   * @param c
   * @return
   */
  protected def popQuestion(c:CmdLineQ) = {
    var d= c.dftAnswer
    var q= c.question
    var ch= c.choices
    var s= ""

    _out.write(q +
        ( if (c.isMust) "*" else "" ) + " ? " )

    if (! STU.isEmpty(ch)) {
      if (ch.indexOf('\n') >= 0) {
        _out.write(
            (if (ch.startsWith("\n")) "[" else "[\n") + ch +
            (if (ch.endsWith("\n")) "]" else "\n]" )
        )
      } else {
        _out.write("[" + ch + "]")
      }
      s= " "
    }
    if (! STU.isEmpty(d)) {
      _out.write("(" + d + ")")
      s= " "
    }

    _out.write(s)
    _out.flush()

    // get the input from user
    s= readData()

    // point to next question, blank ends it
    if (isCanceled) {
      println("")
      ""
    } else {
      c.setAnswer(s)
    }

  }

  private def readData() = {
    var b=new StringBuilder()
    var loop=true
    var esc=false
    var c=0

    // windows has '\r\n'
    // linux has '\n'
    while(loop) {
      c=  _in.read()
      if (c== -1 || c==4) { esc=true; loop=false }
      if (c== '\n') { loop=false }
      if (c=='\r' || c== '\b'|| c==27 /*esc*/) { /* continue */ }
      else if (loop) { b.append(c.toChar) }
    }

    if (esc) {
      _canceled=true ; b.setLength(0)
    }

    STU.trim( b.toString )
  }

  private def end() = onEnd()

  /**
   * @return
   */
  protected def onStart(): String

  /**
   *
   */
  protected def onEnd() {}

  /**
   * @param par
   * @param qs
   */
  def this(par: Option[CmdLineSeq], qs:CmdLineQ*) {
    this(qs: _*)
    _par= par.getOrElse( null )
  }

}

File: src/main/clojure/com/zotoh/frwk/util/cmdlineseq.clj
(ns com.zotoh.frwk.util.cmdlineseq
  (:import (java.io BufferedOutputStream InputStreamReader OutputStreamWriter))
  (:import (java.io Reader Writer))
  (:import (java.util Properties))
  (:import (org.apache.commons.lang3 StringUtils))
  (:require [ com.zotoh.frwk.util.coreutils :as CU])
  (:require [ com.zotoh.frwk.util.strutils :as SU])
  )

(defrecord CmdSeqQ [ qid qline choices dft must onok ])

(defn- readData [cout cin]
  (let [ buf (StringBuilder.)
         ms (loop [ c (.read cin) ]
                  ;; windows has '\r\n' linux has '\n'
                  (let [ m (cond
                              (or (= c -1)(= c 4))  #{ :quit :break }
                              (= c (int \newline)) #{ :break }
                              (or (= c (int \return)) (= c (int \backspace)) (= c 27)) #{}
                              :else (do (.append buf (char c)) #{})) ]
                    (if (clojure.core/contains? m :break)
                      m
                      (recur (.read cin))))) ]
    (if (clojure.core/contains? ms :quit) nil (.trim (.toString buf)))))

(defn- popQQ [cout cin cmdQ props]
  (let [ must (:must cmdQ)
         dft (:dft cmdQ)
         onResp (:onok cmdQ)
         q (:qline cmdQ)
         chs (:choices cmdQ) ]
    (.write cout (str q (if must "*" "" ) " ? "))
    (if-not (StringUtils/isEmpty chs)
      (if (SU/contains? chs \n)
        (do (.write cout (str
              (if (.startsWith chs "\n") "[" "[\n")  chs 
              (if (.endsWith chs "\n") "]" "\n]" ) )))
        (do (.write cout (str "[" chs "]")))))
    (if-not (StringUtils/isEmpty dft)
      (.write cout (str "(" dft ")")) )
    (.write cout " ")
    (.flush cout)
    ;; get the input from user
    ;; point to next question, blank ends it
    (let [ rc (readData cout cin)]
      (if (nil? rc)
        (do (.write cout "\n") nil )
        (do (onResp (if (StringUtils/isEmpty rc) dft rc) props))))))

(defn- popQ [cout cin cmdQ props]
  (if (nil? cmdQ)
    ""
    (popQQ cout cin cmdQ props)))


(defn- cycleQ [cout cin cmdQNs start props]
  (let []
    (loop [ rc (popQ cout cin (get cmdQNs start) props) ]
      (cond
        (StringUtils/isEmpty rc) props
        (nil? rc) nil
        :else (recur (popQ cout cin (get cmdQNs rc) props))))))

(defn start [cmdQs q1]
  (let [ kp (if (CU/isWindows) "<Ctrl-C>" "<Ctrl-D>")
         cout (OutputStreamWriter. (BufferedOutputStream. (System/out)))
         cin (InputStreamReader. (System/in))
         props (Properties.) ]
    (.write cout (str ">>> Press " kp "<Enter> to cancel...\n"))
    (cycleQ cout cin cmdQs q1 props)))


(comment
(def q1 (->CmdSeqQ "q1" "hello ken" "q|b|c" "c" true 
           (fn [a ps]
             (do (.put ps "a1" a) "q2")) ) )
(def q2 (->CmdSeqQ "q2" "hello paul" "" "" false 
           (fn [a ps]
             (do (.put ps "a2" a) "q3"))) )
(def q3 (->CmdSeqQ "q3" "hello joe" "z" "" false 
           (fn [a ps]
             (do (.put ps "a3" a) "" ))) )
(def QM { "q1" q1 "q2" q2 "q3" q3 })
)

(def ^:private cmdlineseq-eof nil)

